{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Phylogenetically balanced database of prokaryotes\n",
    "\n",
    "Constructed as a subset from [GTDB release 214](https://gtdb.ecogenomic.org/).\n",
    "\n",
    "Nickname: `DB_Proka`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "\n",
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_folder = Path('./data')\n",
    "assert data_folder.is_dir()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load GTDB metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genomes: 402,709\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_contamination</th>\n",
       "      <th>coding_bases</th>\n",
       "      <th>coding_density</th>\n",
       "      <th>contig_count</th>\n",
       "      <th>gc_count</th>\n",
       "      <th>gc_percentage</th>\n",
       "      <th>genome_size</th>\n",
       "      <th>...</th>\n",
       "      <th>gtdb_type_species_of_genus</th>\n",
       "      <th>gtdb_taxonomy</th>\n",
       "      <th>ncbi_taxonomy</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>gtdb_representative_b</th>\n",
       "      <th>gtdb_type_species_of_genus_b</th>\n",
       "      <th>n_genomes_in_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RS_GCF_000246985.2</td>\n",
       "      <td>Thermococcus litoralis DSM 5473</td>\n",
       "      <td>99.50</td>\n",
       "      <td>0.50</td>\n",
       "      <td>2014456</td>\n",
       "      <td>90.939033</td>\n",
       "      <td>1</td>\n",
       "      <td>954455</td>\n",
       "      <td>43.088029</td>\n",
       "      <td>2215172</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>d__Archaea;p__Methanobacteriota_B;c__Thermococ...</td>\n",
       "      <td>d__Archaea;p__Euryarchaeota;c__Thermococci;o__...</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__Methanobacteriota_B</td>\n",
       "      <td>g__Thermococcus_A</td>\n",
       "      <td>s__Thermococcus_A alcaliphilus</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RS_GCF_000980135.1</td>\n",
       "      <td>Methanosarcina mazei</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.65</td>\n",
       "      <td>3081035</td>\n",
       "      <td>75.560622</td>\n",
       "      <td>193</td>\n",
       "      <td>1694808</td>\n",
       "      <td>41.570885</td>\n",
       "      <td>4077567</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>d__Archaea;p__Halobacteriota;c__Methanosarcini...</td>\n",
       "      <td>d__Archaea;p__Euryarchaeota;c__Methanomicrobia...</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__Halobacteriota</td>\n",
       "      <td>g__Methanosarcina</td>\n",
       "      <td>s__Methanosarcina mazei</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RS_GCF_000337075.1</td>\n",
       "      <td>Halorubrum hochstenium ATCC 700873</td>\n",
       "      <td>99.38</td>\n",
       "      <td>0.19</td>\n",
       "      <td>2629910</td>\n",
       "      <td>86.580487</td>\n",
       "      <td>64</td>\n",
       "      <td>2099820</td>\n",
       "      <td>69.130537</td>\n",
       "      <td>3037532</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>d__Archaea;p__Halobacteriota;c__Halobacteria;o...</td>\n",
       "      <td>d__Archaea;p__Euryarchaeota;c__Halobacteria;o_...</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__Halobacteriota</td>\n",
       "      <td>g__Halorubrum</td>\n",
       "      <td>s__Halorubrum hochstenium</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RS_GCF_000979515.1</td>\n",
       "      <td>Methanosarcina mazei</td>\n",
       "      <td>100.00</td>\n",
       "      <td>1.31</td>\n",
       "      <td>3083899</td>\n",
       "      <td>75.657950</td>\n",
       "      <td>215</td>\n",
       "      <td>1696947</td>\n",
       "      <td>41.632859</td>\n",
       "      <td>4076107</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>d__Archaea;p__Halobacteriota;c__Methanosarcini...</td>\n",
       "      <td>d__Archaea;p__Euryarchaeota;c__Methanomicrobia...</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__Halobacteriota</td>\n",
       "      <td>g__Methanosarcina</td>\n",
       "      <td>s__Methanosarcina mazei</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RS_GCF_000762265.1</td>\n",
       "      <td>Methanobacterium formicicum</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2045404</td>\n",
       "      <td>83.486321</td>\n",
       "      <td>1</td>\n",
       "      <td>1012813</td>\n",
       "      <td>41.339525</td>\n",
       "      <td>2449987</td>\n",
       "      <td>...</td>\n",
       "      <td>f</td>\n",
       "      <td>d__Archaea;p__Methanobacteriota;c__Methanobact...</td>\n",
       "      <td>d__Archaea;p__Euryarchaeota;c__Methanobacteria...</td>\n",
       "      <td>d__Archaea</td>\n",
       "      <td>p__Methanobacteriota</td>\n",
       "      <td>g__Methanobacterium</td>\n",
       "      <td>s__Methanobacterium formicicum</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            accession                  ncbi_organism_name  \\\n",
       "0  RS_GCF_000246985.2     Thermococcus litoralis DSM 5473   \n",
       "1  RS_GCF_000980135.1                Methanosarcina mazei   \n",
       "2  RS_GCF_000337075.1  Halorubrum hochstenium ATCC 700873   \n",
       "3  RS_GCF_000979515.1                Methanosarcina mazei   \n",
       "4  RS_GCF_000762265.1         Methanobacterium formicicum   \n",
       "\n",
       "   checkm_completeness  checkm_contamination  coding_bases  coding_density  \\\n",
       "0                99.50                  0.50       2014456       90.939033   \n",
       "1               100.00                  0.65       3081035       75.560622   \n",
       "2                99.38                  0.19       2629910       86.580487   \n",
       "3               100.00                  1.31       3083899       75.657950   \n",
       "4               100.00                  0.00       2045404       83.486321   \n",
       "\n",
       "   contig_count  gc_count  gc_percentage  genome_size  ...  \\\n",
       "0             1    954455      43.088029      2215172  ...   \n",
       "1           193   1694808      41.570885      4077567  ...   \n",
       "2            64   2099820      69.130537      3037532  ...   \n",
       "3           215   1696947      41.632859      4076107  ...   \n",
       "4             1   1012813      41.339525      2449987  ...   \n",
       "\n",
       "   gtdb_type_species_of_genus  \\\n",
       "0                           f   \n",
       "1                           f   \n",
       "2                           f   \n",
       "3                           f   \n",
       "4                           f   \n",
       "\n",
       "                                       gtdb_taxonomy  \\\n",
       "0  d__Archaea;p__Methanobacteriota_B;c__Thermococ...   \n",
       "1  d__Archaea;p__Halobacteriota;c__Methanosarcini...   \n",
       "2  d__Archaea;p__Halobacteriota;c__Halobacteria;o...   \n",
       "3  d__Archaea;p__Halobacteriota;c__Methanosarcini...   \n",
       "4  d__Archaea;p__Methanobacteriota;c__Methanobact...   \n",
       "\n",
       "                                       ncbi_taxonomy      domain  \\\n",
       "0  d__Archaea;p__Euryarchaeota;c__Thermococci;o__...  d__Archaea   \n",
       "1  d__Archaea;p__Euryarchaeota;c__Methanomicrobia...  d__Archaea   \n",
       "2  d__Archaea;p__Euryarchaeota;c__Halobacteria;o_...  d__Archaea   \n",
       "3  d__Archaea;p__Euryarchaeota;c__Methanomicrobia...  d__Archaea   \n",
       "4  d__Archaea;p__Euryarchaeota;c__Methanobacteria...  d__Archaea   \n",
       "\n",
       "                   phylum                genus  \\\n",
       "0  p__Methanobacteriota_B    g__Thermococcus_A   \n",
       "1       p__Halobacteriota    g__Methanosarcina   \n",
       "2       p__Halobacteriota        g__Halorubrum   \n",
       "3       p__Halobacteriota    g__Methanosarcina   \n",
       "4    p__Methanobacteriota  g__Methanobacterium   \n",
       "\n",
       "                          species gtdb_representative_b  \\\n",
       "0  s__Thermococcus_A alcaliphilus                 False   \n",
       "1         s__Methanosarcina mazei                 False   \n",
       "2       s__Halorubrum hochstenium                  True   \n",
       "3         s__Methanosarcina mazei                 False   \n",
       "4  s__Methanobacterium formicicum                 False   \n",
       "\n",
       "  gtdb_type_species_of_genus_b  n_genomes_in_species  \n",
       "0                        False                     3  \n",
       "1                        False                    78  \n",
       "2                        False                     2  \n",
       "3                        False                    78  \n",
       "4                        False                    10  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = [\n",
    "    'accession', 'ncbi_organism_name', 'checkm_completeness', 'checkm_contamination', 'coding_bases', 'coding_density', \n",
    "    'contig_count', 'gc_count', 'gc_percentage', 'genome_size', 'protein_count', 'gtdb_representative', \n",
    "    'gtdb_type_species_of_genus', 'gtdb_taxonomy', 'ncbi_taxonomy', \n",
    "]\n",
    "\n",
    "def extract_taxonomy(tax, tax_one_letter_code):\n",
    "    for t in tax.split(';'):\n",
    "        if t.startswith(f'{tax_one_letter_code}__'):\n",
    "            return t\n",
    "        \n",
    "    \n",
    "gtdb_metadata = pd.concat(\n",
    "    [\n",
    "        pd.read_csv(\n",
    "            data_folder / 'gtdb_r214.1' / 'ar53_metadata_r214.tsv.gz', \n",
    "            sep='\\t', \n",
    "            usecols=columns,\n",
    "        )[columns].reset_index(drop=True),\n",
    "        pd.read_csv(\n",
    "            data_folder / 'gtdb_r214.1' / 'bac120_metadata_r214.tsv.gz', \n",
    "            sep='\\t', \n",
    "            usecols=columns,\n",
    "        )[columns].reset_index(drop=True)\n",
    "    ],\n",
    "    ignore_index=True,\n",
    ")\n",
    "\n",
    "gtdb_metadata['domain'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: extract_taxonomy(t, 'd'))\n",
    "gtdb_metadata['phylum'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: extract_taxonomy(t, 'p'))\n",
    "gtdb_metadata['genus'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: extract_taxonomy(t, 'g'))\n",
    "gtdb_metadata['species'] = gtdb_metadata['gtdb_taxonomy'].apply(lambda t: extract_taxonomy(t, 's'))\n",
    "gtdb_metadata['gtdb_representative_b'] = gtdb_metadata['gtdb_representative'].apply(lambda v: v == 't')\n",
    "gtdb_metadata['gtdb_type_species_of_genus_b'] = gtdb_metadata['gtdb_type_species_of_genus'].apply(lambda v: v == 't')\n",
    "\n",
    "species_stats = gtdb_metadata[['species', 'accession']].groupby('species').count()\n",
    "n_genomes_in_species = {\n",
    "    species: species_stats.loc[species]['accession'] for species in species_stats.index\n",
    "}\n",
    "gtdb_metadata['n_genomes_in_species'] = gtdb_metadata['species'].apply(lambda s: n_genomes_in_species[s])\n",
    "\n",
    "print(f'Number of genomes: {len(gtdb_metadata):,}')\n",
    "gtdb_metadata.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select genomes\n",
    "\n",
    "- All genera\n",
    "- Up to 10 species per genus\n",
    "    - \"type\" species are selected in priority\n",
    "    - otherwise, species with more strains are selected in priority\n",
    "    - tiebreak: genome completeness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genus 1 / 20,739\n",
      "Genus 10,000 / 20,739\n",
      "Genus 20,000 / 20,739\n",
      "Genus 20,739 / 20,739\n"
     ]
    }
   ],
   "source": [
    "def make_selection(gtdb_metadata, n_species_per_genus=10):\n",
    "    accession_list = []\n",
    "\n",
    "    gtdb_metadata_sorted = gtdb_metadata.sort_values(\n",
    "        [\n",
    "            'genus', \n",
    "            'gtdb_type_species_of_genus_b', \n",
    "            'gtdb_representative_b',\n",
    "            'n_genomes_in_species',\n",
    "            'checkm_completeness',\n",
    "            'checkm_contamination',\n",
    "            'protein_count',\n",
    "            'accession',\n",
    "        ], \n",
    "        ascending=[\n",
    "            True,  # genus\n",
    "            False, # gtdb_type_species_of_genus_b: true first\n",
    "            False, # gtdb_representative_b: true first\n",
    "            False, # n_genomes_in_species: high number of genomes first\n",
    "            False, # checkm_completeness: high completness first\n",
    "            True,  # checkm_contamination: small contaminations first\n",
    "            False, # protein_count: large genomes first\n",
    "            True,  # accession: if else fails sort on accession\n",
    "        ]\n",
    "    ).reset_index(drop=True)\n",
    "\n",
    "    # We make one exception for Halogranum salarium B-1 (Halogranum rubrum in GTDB).\n",
    "    # The selected strain for this species is not the one we studied prior to building this DB.\n",
    "    # We manually select the strain that we want included instead.\n",
    "    hgr_rubrum_accession = 'RS_GCF_000283335.1'\n",
    "\n",
    "    genera = gtdb_metadata_sorted['genus'].unique()\n",
    "    for i, genus in enumerate(genera):\n",
    "        if i == 0 or (i+1) % 10000 == 0 or (i+1) == len(genera):\n",
    "            print(f'Genus {i+1:,} / {len(genera):,}')\n",
    "\n",
    "        genus_df_sorted = gtdb_metadata_sorted[\n",
    "            gtdb_metadata_sorted['genus'] == genus\n",
    "        ].reset_index(drop=True)\n",
    "\n",
    "        selection_df = genus_df_sorted.drop_duplicates(\n",
    "            ['genus', 'species']\n",
    "        ).head(n_species_per_genus)\n",
    "\n",
    "        if genus == 'g__Halogranum':\n",
    "            l = selection_df[\n",
    "                selection_df['species'] != 's__Halogranum rubrum'\n",
    "            ]['accession'].values.tolist()\n",
    "            l.append(hgr_rubrum_accession)\n",
    "            assert len(l) == 3\n",
    "        else:\n",
    "            l = selection_df['accession'].values.tolist()\n",
    "        \n",
    "        accession_list.extend(l)\n",
    "\n",
    "    assert hgr_rubrum_accession in accession_list\n",
    "\n",
    "    return accession_list\n",
    "        \n",
    "\n",
    "accessions = make_selection(gtdb_metadata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "domain\n",
       "d__Bacteria    50713\n",
       "d__Archaea      3708\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtdb_metadata_shortlist = gtdb_metadata[gtdb_metadata['accession'].isin(accessions)].reset_index(drop=True)\n",
    "gtdb_metadata_shortlist['domain'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some of the accessions in GTDB release 214 did not exist on NCBI anymore, i.e. the final number of species is slightly down from the above:\n",
    "\n",
    "- Archaea:   3,706\n",
    "- Bacteria: 50,640"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Downloading genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genomes were downloaded and processed using this package: https://github.com/srom/assembly\n",
    "(created as part of this work, too)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AvB",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
