{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Figure 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from Bio import Phylo\n",
    "\n",
    "cwd = os.getcwd()\n",
    "if cwd.endswith('notebook'):\n",
    "    os.chdir('..')\n",
    "    cwd = os.getcwd()\n",
    "\n",
    "from src.tree.itol_annotation import itol_heatmap_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_palette('colorblind')\n",
    "sns.set_style('whitegrid')\n",
    "sns.set_context('paper', font_scale=1.8)\n",
    "plt.rcParams['font.family'] = 'Helvetica'\n",
    "\n",
    "palette = sns.color_palette().as_hex()\n",
    "\n",
    "data_folder = Path('./data')\n",
    "assert data_folder.is_dir()\n",
    "\n",
    "figures_folder = Path('./figures')\n",
    "assert figures_folder.is_dir()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdb_metadata = pd.read_csv(data_folder / 'gtdb_metadata.csv', index_col='ncbi_accession')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bacteriocins heatmap\n",
    "\n",
    "Data generated with script `src/TBD`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>amp_id</th>\n",
       "      <th>amp_desc</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>ncbi_accession</th>\n",
       "      <th>domain</th>\n",
       "      <th>gtdb_phylum</th>\n",
       "      <th>gtdb_class</th>\n",
       "      <th>gtdb_order</th>\n",
       "      <th>gtdb_family</th>\n",
       "      <th>gtdb_genus</th>\n",
       "      <th>gtdb_species</th>\n",
       "      <th>ncbi_organism_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000007325.1</td>\n",
       "      <td>AAL94370.1</td>\n",
       "      <td>pg_hydrolase</td>\n",
       "      <td>Amidase_2+PG_binding_1</td>\n",
       "      <td>1.150000e-08</td>\n",
       "      <td>61.95</td>\n",
       "      <td>GCA_000007325.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Fusobacteriota</td>\n",
       "      <td>Fusobacteriia</td>\n",
       "      <td>Fusobacteriales</td>\n",
       "      <td>Fusobacteriaceae</td>\n",
       "      <td>Fusobacterium</td>\n",
       "      <td>Fusobacterium nucleatum</td>\n",
       "      <td>Fusobacterium nucleatum subsp. nucleatum ATCC ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000007325.1</td>\n",
       "      <td>AAL94673.1</td>\n",
       "      <td>pg_hydrolase</td>\n",
       "      <td>Peptidase_M23+LysM</td>\n",
       "      <td>6.500000e-08</td>\n",
       "      <td>64.00</td>\n",
       "      <td>GCA_000007325.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Fusobacteriota</td>\n",
       "      <td>Fusobacteriia</td>\n",
       "      <td>Fusobacteriales</td>\n",
       "      <td>Fusobacteriaceae</td>\n",
       "      <td>Fusobacterium</td>\n",
       "      <td>Fusobacterium nucleatum</td>\n",
       "      <td>Fusobacterium nucleatum subsp. nucleatum ATCC ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000008885.1</td>\n",
       "      <td>BAC24414.1</td>\n",
       "      <td>pg_hydrolase</td>\n",
       "      <td>Amidase_3+AMIN</td>\n",
       "      <td>3.900000e-15</td>\n",
       "      <td>106.70</td>\n",
       "      <td>GCA_000008885.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonadota</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Wigglesworthia</td>\n",
       "      <td>Wigglesworthia glossinidia_A</td>\n",
       "      <td>Wigglesworthia glossinidia endosymbiont of Glo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000010565.1</td>\n",
       "      <td>BAF58967.1</td>\n",
       "      <td>pg_hydrolase</td>\n",
       "      <td>Amidase_3+AMIN</td>\n",
       "      <td>6.000000e-15</td>\n",
       "      <td>121.55</td>\n",
       "      <td>GCA_000010565.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota_B</td>\n",
       "      <td>Desulfotomaculia</td>\n",
       "      <td>Desulfotomaculales</td>\n",
       "      <td>Pelotomaculaceae</td>\n",
       "      <td>Pelotomaculum</td>\n",
       "      <td>Pelotomaculum thermopropionicum</td>\n",
       "      <td>Pelotomaculum thermopropionicum SI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000010565.1</td>\n",
       "      <td>BAF60354.1</td>\n",
       "      <td>pg_hydrolase</td>\n",
       "      <td>Amidase_2+LysM</td>\n",
       "      <td>6.000135e-15</td>\n",
       "      <td>59.45</td>\n",
       "      <td>GCA_000010565.1</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacillota_B</td>\n",
       "      <td>Desulfotomaculia</td>\n",
       "      <td>Desulfotomaculales</td>\n",
       "      <td>Pelotomaculaceae</td>\n",
       "      <td>Pelotomaculum</td>\n",
       "      <td>Pelotomaculum thermopropionicum</td>\n",
       "      <td>Pelotomaculum thermopropionicum SI</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly_accession  protein_id        amp_id                amp_desc  \\\n",
       "0    GCA_000007325.1  AAL94370.1  pg_hydrolase  Amidase_2+PG_binding_1   \n",
       "1    GCA_000007325.1  AAL94673.1  pg_hydrolase      Peptidase_M23+LysM   \n",
       "2    GCA_000008885.1  BAC24414.1  pg_hydrolase          Amidase_3+AMIN   \n",
       "3    GCA_000010565.1  BAF58967.1  pg_hydrolase          Amidase_3+AMIN   \n",
       "4    GCA_000010565.1  BAF60354.1  pg_hydrolase          Amidase_2+LysM   \n",
       "\n",
       "         evalue  bitscore   ncbi_accession    domain     gtdb_phylum  \\\n",
       "0  1.150000e-08     61.95  GCA_000007325.1  Bacteria  Fusobacteriota   \n",
       "1  6.500000e-08     64.00  GCA_000007325.1  Bacteria  Fusobacteriota   \n",
       "2  3.900000e-15    106.70  GCA_000008885.1  Bacteria  Pseudomonadota   \n",
       "3  6.000000e-15    121.55  GCA_000010565.1  Bacteria     Bacillota_B   \n",
       "4  6.000135e-15     59.45  GCA_000010565.1  Bacteria     Bacillota_B   \n",
       "\n",
       "            gtdb_class          gtdb_order         gtdb_family  \\\n",
       "0        Fusobacteriia     Fusobacteriales    Fusobacteriaceae   \n",
       "1        Fusobacteriia     Fusobacteriales    Fusobacteriaceae   \n",
       "2  Gammaproteobacteria    Enterobacterales  Enterobacteriaceae   \n",
       "3     Desulfotomaculia  Desulfotomaculales    Pelotomaculaceae   \n",
       "4     Desulfotomaculia  Desulfotomaculales    Pelotomaculaceae   \n",
       "\n",
       "       gtdb_genus                     gtdb_species  \\\n",
       "0   Fusobacterium          Fusobacterium nucleatum   \n",
       "1   Fusobacterium          Fusobacterium nucleatum   \n",
       "2  Wigglesworthia     Wigglesworthia glossinidia_A   \n",
       "3   Pelotomaculum  Pelotomaculum thermopropionicum   \n",
       "4   Pelotomaculum  Pelotomaculum thermopropionicum   \n",
       "\n",
       "                                  ncbi_organism_name  \n",
       "0  Fusobacterium nucleatum subsp. nucleatum ATCC ...  \n",
       "1  Fusobacterium nucleatum subsp. nucleatum ATCC ...  \n",
       "2  Wigglesworthia glossinidia endosymbiont of Glo...  \n",
       "3                 Pelotomaculum thermopropionicum SI  \n",
       "4                 Pelotomaculum thermopropionicum SI  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriocins_df = pd.read_csv(data_folder / 'bacteriocins.csv.gz')\n",
    "bacteriocins_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make sure the bacteriocins dataset is congruent with the PGH dataset "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "pgh_df = pd.read_csv(data_folder / 'pgh_proteins.csv')\n",
    "pgh_ids = set(pgh_df['protein_id'].unique())\n",
    "pgh_bacteriocins = bacteriocins_df[bacteriocins_df['amp_id'] == 'pg_hydrolase']\n",
    "pgh_bacteriocins_ids = set(pgh_bacteriocins['protein_id'].unique())\n",
    "\n",
    "assert len(pgh_df) == len(pgh_bacteriocins)\n",
    "assert len(pgh_bacteriocins_ids - pgh_ids) == 0\n",
    "assert len(pgh_ids - pgh_bacteriocins_ids) == 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assign phylum group\n",
    "\n",
    "Just like in figre 1 â€“ GTDB phyla sub group such as Pseudomonata_A, Pseudomonata_B, etc. are merged into Pseudomonata for clarity. Unamed phyla are ignored."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_named_phylum(phylum):\n",
    "    if re.match(r'^[A-Z0-9\\-_]+$', phylum) is not None:\n",
    "        return False\n",
    "\n",
    "    blacklist = {\n",
    "        'BMS3Abin14',\n",
    "        'BS750m-G25',\n",
    "        'BS750m-G34',\n",
    "        'JdFR-76',\n",
    "        'T1Sed10-126',\n",
    "        'SpSt-1190',\n",
    "        'B1Sed10-29',\n",
    "        'EX4484-52',\n",
    "    }\n",
    "    return phylum not in blacklist\n",
    "\n",
    "\n",
    "def get_phylum_group(phylum):\n",
    "    m = re.match(r'^(.+)_[A-Z]$', phylum)\n",
    "    if m is not None:\n",
    "        return m[1]\n",
    "    else:\n",
    "        return phylum\n",
    "\n",
    "\n",
    "named_phyla = [\n",
    "    p for p in gtdb_metadata['gtdb_phylum'].unique()\n",
    "    if is_named_phylum(p)\n",
    "]\n",
    "map_phylm_groups = {\n",
    "    phylum: get_phylum_group(phylum)\n",
    "    for phylum in named_phyla\n",
    "}\n",
    "bacteriocins_df['phylum_group'] = bacteriocins_df['gtdb_phylum'].apply(lambda p: map_phylm_groups.get(p))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of archaeal phylum groups: 14\n",
      "Number of bacterial phylum groups: 34\n"
     ]
    }
   ],
   "source": [
    "arc_phylum_group_tree = Phylo.read(data_folder / 'figure1' / 'ar53_phylum_groups.tree', 'phyloxml')\n",
    "bac_phylum_group_tree = Phylo.read(data_folder / 'figure1' / 'bac120_phylum_groups.tree', 'phyloxml')\n",
    "\n",
    "arc_phylum_groups = sorted([l.name for l in arc_phylum_group_tree.get_terminals()])\n",
    "bac_phylum_groups = sorted([l.name for l in bac_phylum_group_tree.get_terminals()])\n",
    "all_phylum_groups = sorted(set(arc_phylum_groups) | set(bac_phylum_groups))\n",
    "\n",
    "print(f'Number of archaeal phylum groups: {len(arc_phylum_groups):,}')\n",
    "print(f'Number of bacterial phylum groups: {len(bac_phylum_groups):,}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of bacteriocins included within phyla subset: 145,436 (98% of total)\n"
     ]
    }
   ],
   "source": [
    "bacteriocins_subset = bacteriocins_df[\n",
    "    bacteriocins_df['phylum_group'].isin(all_phylum_groups)\n",
    "].copy()\n",
    "p = 100 * len(bacteriocins_subset) / len(bacteriocins_df)\n",
    "\n",
    "print(f'Number of bacteriocins included within phyla subset: {len(bacteriocins_subset):,} ({p:.0f}% of total)')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selection of bacteriocins shared between archaea and bacteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of bacteriocins shared between archaea and bacteria: 86\n"
     ]
    }
   ],
   "source": [
    "arc_bacteriocins = set(bacteriocins_subset[bacteriocins_subset['domain'] == 'Archaea']['amp_id'].unique())\n",
    "bac_bacteriocins = set(bacteriocins_subset[bacteriocins_subset['domain'] == 'Bacteria']['amp_id'].unique())\n",
    "\n",
    "shared_bacteriocins = sorted(arc_bacteriocins & bac_bacteriocins)\n",
    "\n",
    "print(f'Number of bacteriocins shared between archaea and bacteria: {len(shared_bacteriocins):,}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_assemblies</th>\n",
       "      <th>n_archaea</th>\n",
       "      <th>n_bacteria</th>\n",
       "      <th>percent_archaea</th>\n",
       "      <th>percent_bacteria</th>\n",
       "      <th>percent_diff</th>\n",
       "      <th>n_archaeal_phyla</th>\n",
       "      <th>n_bacterial_phyla</th>\n",
       "      <th>n_archaeal_classes</th>\n",
       "      <th>n_bacterial_classes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>amp_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pg_hydrolase</th>\n",
       "      <td>41535</td>\n",
       "      <td>174</td>\n",
       "      <td>41361</td>\n",
       "      <td>4.70</td>\n",
       "      <td>81.68</td>\n",
       "      <td>76.98</td>\n",
       "      <td>13</td>\n",
       "      <td>169</td>\n",
       "      <td>23</td>\n",
       "      <td>459</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Linocin_M18</th>\n",
       "      <td>2559</td>\n",
       "      <td>174</td>\n",
       "      <td>2385</td>\n",
       "      <td>4.70</td>\n",
       "      <td>4.71</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9</td>\n",
       "      <td>67</td>\n",
       "      <td>21</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FlvA2h</th>\n",
       "      <td>13</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-0.15</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactococcin_A</th>\n",
       "      <td>149</td>\n",
       "      <td>5</td>\n",
       "      <td>144</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.15</td>\n",
       "      <td>4</td>\n",
       "      <td>26</td>\n",
       "      <td>5</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nocardithiocin</th>\n",
       "      <td>43</td>\n",
       "      <td>5</td>\n",
       "      <td>38</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.08</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ComX4</th>\n",
       "      <td>52</td>\n",
       "      <td>4</td>\n",
       "      <td>48</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.09</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Competence</th>\n",
       "      <td>39</td>\n",
       "      <td>4</td>\n",
       "      <td>35</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.07</td>\n",
       "      <td>-0.04</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactococcin_Q_chain_b</th>\n",
       "      <td>26</td>\n",
       "      <td>4</td>\n",
       "      <td>22</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.04</td>\n",
       "      <td>-0.07</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Enterocin_NKR_5_3D</th>\n",
       "      <td>28</td>\n",
       "      <td>4</td>\n",
       "      <td>24</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.05</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>4</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactococcin_MMFII</th>\n",
       "      <td>12</td>\n",
       "      <td>4</td>\n",
       "      <td>8</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sakacin_A</th>\n",
       "      <td>13</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Auto_Inducing_Peptide_III</th>\n",
       "      <td>15</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FlvA2f</th>\n",
       "      <td>12</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>EnterocinP</th>\n",
       "      <td>26</td>\n",
       "      <td>3</td>\n",
       "      <td>23</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.05</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Garvicin_Q</th>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteriocin_31</th>\n",
       "      <td>40</td>\n",
       "      <td>2</td>\n",
       "      <td>38</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.03</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Enterocin_1071B</th>\n",
       "      <td>67</td>\n",
       "      <td>2</td>\n",
       "      <td>65</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.13</td>\n",
       "      <td>0.08</td>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ComX2</th>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PaeM</th>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>16</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Piricyclamide_7005E2</th>\n",
       "      <td>33</td>\n",
       "      <td>2</td>\n",
       "      <td>31</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.01</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Caulonodin_III</th>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.02</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gardimycin_(actagardine)</th>\n",
       "      <td>15</td>\n",
       "      <td>2</td>\n",
       "      <td>13</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-0.02</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           n_assemblies  n_archaea  n_bacteria  \\\n",
       "amp_id                                                           \n",
       "pg_hydrolase                      41535        174       41361   \n",
       "Linocin_M18                        2559        174        2385   \n",
       "FlvA2h                               13          6           7   \n",
       "Lactococcin_A                       149          5         144   \n",
       "Nocardithiocin                       43          5          38   \n",
       "ComX4                                52          4          48   \n",
       "Competence                           39          4          35   \n",
       "Lactococcin_Q_chain_b                26          4          22   \n",
       "Enterocin_NKR_5_3D                   28          4          24   \n",
       "Lactococcin_MMFII                    12          4           8   \n",
       "Sakacin_A                            13          3          10   \n",
       "Auto_Inducing_Peptide_III            15          3          12   \n",
       "FlvA2f                               12          3           9   \n",
       "EnterocinP                           26          3          23   \n",
       "Garvicin_Q                           10          2           8   \n",
       "Bacteriocin_31                       40          2          38   \n",
       "Enterocin_1071B                      67          2          65   \n",
       "ComX2                                12          2          10   \n",
       "PaeM                                 18          2          16   \n",
       "Piricyclamide_7005E2                 33          2          31   \n",
       "Caulonodin_III                       12          2          10   \n",
       "Gardimycin_(actagardine)             15          2          13   \n",
       "\n",
       "                           percent_archaea  percent_bacteria  percent_diff  \\\n",
       "amp_id                                                                       \n",
       "pg_hydrolase                          4.70             81.68         76.98   \n",
       "Linocin_M18                           4.70              4.71          0.01   \n",
       "FlvA2h                                0.16              0.01         -0.15   \n",
       "Lactococcin_A                         0.13              0.28          0.15   \n",
       "Nocardithiocin                        0.13              0.08         -0.05   \n",
       "ComX4                                 0.11              0.09         -0.02   \n",
       "Competence                            0.11              0.07         -0.04   \n",
       "Lactococcin_Q_chain_b                 0.11              0.04         -0.07   \n",
       "Enterocin_NKR_5_3D                    0.11              0.05         -0.06   \n",
       "Lactococcin_MMFII                     0.11              0.02         -0.09   \n",
       "Sakacin_A                             0.08              0.02         -0.06   \n",
       "Auto_Inducing_Peptide_III             0.08              0.02         -0.06   \n",
       "FlvA2f                                0.08              0.02         -0.06   \n",
       "EnterocinP                            0.08              0.05         -0.03   \n",
       "Garvicin_Q                            0.05              0.02         -0.03   \n",
       "Bacteriocin_31                        0.05              0.08          0.03   \n",
       "Enterocin_1071B                       0.05              0.13          0.08   \n",
       "ComX2                                 0.05              0.02         -0.03   \n",
       "PaeM                                  0.05              0.03         -0.02   \n",
       "Piricyclamide_7005E2                  0.05              0.06          0.01   \n",
       "Caulonodin_III                        0.05              0.02         -0.03   \n",
       "Gardimycin_(actagardine)              0.05              0.03         -0.02   \n",
       "\n",
       "                           n_archaeal_phyla  n_bacterial_phyla  \\\n",
       "amp_id                                                           \n",
       "pg_hydrolase                             13                169   \n",
       "Linocin_M18                               9                 67   \n",
       "FlvA2h                                    4                  6   \n",
       "Lactococcin_A                             4                 26   \n",
       "Nocardithiocin                            2                 11   \n",
       "ComX4                                     3                 12   \n",
       "Competence                                4                  8   \n",
       "Lactococcin_Q_chain_b                     2                 13   \n",
       "Enterocin_NKR_5_3D                        4                 10   \n",
       "Lactococcin_MMFII                         3                  5   \n",
       "Sakacin_A                                 3                  9   \n",
       "Auto_Inducing_Peptide_III                 2                 10   \n",
       "FlvA2f                                    3                  8   \n",
       "EnterocinP                                3                  9   \n",
       "Garvicin_Q                                2                  5   \n",
       "Bacteriocin_31                            2                 11   \n",
       "Enterocin_1071B                           2                 19   \n",
       "ComX2                                     3                  7   \n",
       "PaeM                                      2                  5   \n",
       "Piricyclamide_7005E2                      2                 13   \n",
       "Caulonodin_III                            2                  6   \n",
       "Gardimycin_(actagardine)                  2                  4   \n",
       "\n",
       "                           n_archaeal_classes  n_bacterial_classes  \n",
       "amp_id                                                              \n",
       "pg_hydrolase                               23                  459  \n",
       "Linocin_M18                                21                  131  \n",
       "FlvA2h                                      4                    7  \n",
       "Lactococcin_A                               5                   42  \n",
       "Nocardithiocin                              2                   16  \n",
       "ComX4                                       3                   14  \n",
       "Competence                                  4                   10  \n",
       "Lactococcin_Q_chain_b                       2                   15  \n",
       "Enterocin_NKR_5_3D                          4                   14  \n",
       "Lactococcin_MMFII                           3                    6  \n",
       "Sakacin_A                                   3                   10  \n",
       "Auto_Inducing_Peptide_III                   2                   11  \n",
       "FlvA2f                                      3                    8  \n",
       "EnterocinP                                  3                   13  \n",
       "Garvicin_Q                                  2                    5  \n",
       "Bacteriocin_31                              2                   14  \n",
       "Enterocin_1071B                             2                   27  \n",
       "ComX2                                       3                    7  \n",
       "PaeM                                        2                    7  \n",
       "Piricyclamide_7005E2                        2                   17  \n",
       "Caulonodin_III                              2                    7  \n",
       "Gardimycin_(actagardine)                    2                    5  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_per_amp = bacteriocins_subset[\n",
    "    bacteriocins_subset['amp_id'] .isin(shared_bacteriocins)\n",
    "][\n",
    "    ['amp_id', 'assembly_accession'\n",
    "]].groupby('amp_id').nunique().rename(columns={'assembly_accession': 'n_assemblies'})\n",
    "\n",
    "counts_per_domain = bacteriocins_subset[\n",
    "    ['domain', 'amp_id', 'assembly_accession'\n",
    "]].groupby(['domain', 'amp_id']).nunique()\n",
    "\n",
    "archaea_counts = counts_per_domain.loc['Archaea'].copy().rename(columns={'assembly_accession': 'n_archaea'})\n",
    "bacteria_counts = counts_per_domain.loc['Bacteria'].copy().rename(columns={'assembly_accession': 'n_bacteria'})\n",
    "\n",
    "def compute_clade_count(df, amp_id, domain, clade='gtdb_class'):\n",
    "    return len(df[\n",
    "        (df['domain'] == domain) &\n",
    "        (df['amp_id'] == amp_id)\n",
    "    ][clade].unique())\n",
    "\n",
    "amp_table = pd.merge(\n",
    "    counts_per_amp,\n",
    "    archaea_counts,\n",
    "    how='left',\n",
    "    on='amp_id',\n",
    ")\n",
    "amp_table = pd.merge(\n",
    "    amp_table,\n",
    "    bacteria_counts,\n",
    "    how='left',\n",
    "    on='amp_id',\n",
    ")\n",
    "amp_table['percent_archaea'] = (\n",
    "    100 * amp_table['n_archaea'] / len(gtdb_metadata[gtdb_metadata['domain'] == 'Archaea'])\n",
    ").round(2)\n",
    "amp_table['percent_bacteria'] = (\n",
    "    100 * amp_table['n_bacteria'] / len(gtdb_metadata[gtdb_metadata['domain'] == 'Bacteria'])\n",
    ").round(2)\n",
    "amp_table['percent_diff'] = (amp_table['percent_bacteria'] - amp_table['percent_archaea']).round(2)\n",
    "amp_table['n_archaeal_phyla'] = [\n",
    "    compute_clade_count(bacteriocins_df, amp_id, domain='Archaea', clade='gtdb_phylum')\n",
    "    for amp_id in amp_table.index\n",
    "]\n",
    "amp_table['n_bacterial_phyla'] = [\n",
    "    compute_clade_count(bacteriocins_df, amp_id, domain='Bacteria', clade='gtdb_phylum')\n",
    "    for amp_id in amp_table.index\n",
    "]\n",
    "amp_table['n_archaeal_classes'] = [\n",
    "    compute_clade_count(bacteriocins_df, amp_id, domain='Archaea', clade='gtdb_class')\n",
    "    for amp_id in amp_table.index\n",
    "]\n",
    "amp_table['n_bacterial_classes'] = [\n",
    "    compute_clade_count(bacteriocins_df, amp_id, domain='Bacteria', clade='gtdb_class')\n",
    "    for amp_id in amp_table.index\n",
    "]\n",
    "\n",
    "# Selection criteria: at least 2 genomes and 2 distinct phyla in each domain.\n",
    "amp_table = amp_table[\n",
    "    (amp_table['n_archaea'] >= 2) & \n",
    "    (amp_table['n_archaeal_phyla'] >= 2) & \n",
    "    (amp_table['n_bacteria'] >= 2) &\n",
    "    (amp_table['n_bacterial_phyla'] >= 2)\n",
    "].copy()\n",
    "\n",
    "amp_table = amp_table.sort_values(['n_archaea'], ascending=False)\n",
    "\n",
    "amp_table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Putative bacteriocins excluded from the heatmap:\n",
    "\n",
    "- `Linocin M18` is shown to have genuine use for harvesting iron in Pyrococcus furiosus: https://www.uniprot.org/uniprotkb/Q8U1L4/entry\n",
    "- `FlvA2h` and `FlvA2f` are part of a `Lantipeptide` in _Ruminococcus flavefaciens_ but these two genes are not the active part as per: https://www.uniprot.org/uniprotkb/P0DQM0/entry\n",
    "\n",
    "Interesting but not sure if we should include:\n",
    "- `Competence`, `ComX2` and `ComX4` are both `comX` in _B. subtilis_. mimic of comX can be used to mess with quorum sensing: https://doi.org/10.1016/j.pbi.2004.05.008\n",
    "- Same story but less clearcut for `Auto_Inducing_Peptide_III` which map to TIGR entry TIGR04223: https://www.ncbi.nlm.nih.gov/genome/annotation_prok/evidence/TIGR04223/\n",
    "\n",
    "Name change:\n",
    "- `PaeM` is better known as `Colicin M` as per automatic Pfam annotations: https://www.uniprot.org/uniprotkb/Q88A25/entry\n",
    "- `Bacteriocin_31` is Pfam domain `Bacteriocin_II`\n",
    "- `Gardimycin_(actagardine)` renamed `Actagardine`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "display_name_map = {\n",
    "    'Competence': 'ComX homolog',\n",
    "    'ComX2': 'ComX homolog',\n",
    "    'ComX4': 'ComX homolog',\n",
    "    'pg_hydrolase': 'Peptidoglycan hydrolase',\n",
    "    'PaeM': 'Colicin M',\n",
    "    'Bacteriocin_31': 'Bacteriocin II',\n",
    "    'Gardimycin_(actagardine)': 'Actagardine',\n",
    "    'Lactococcin_Q_chain_b': 'Lactococcin Q',\n",
    "    'Enterocin_NKR_5_3D': 'Enterocin NKR-5-3D',\n",
    "    'EnterocinP': 'Enterocin P',\n",
    "    'Subtilosin_(SboX)': 'Subtilosin',\n",
    "    'GarvieacinQ garQ': 'Garvieacin Q',\n",
    "}\n",
    "def set_display_name(name):\n",
    "    if name in display_name_map:\n",
    "        return display_name_map[name]\n",
    "    else:\n",
    "        return name.replace('_', ' ')\n",
    "    \n",
    "bacteriocins_df['display_name'] = bacteriocins_df['amp_id'].apply(set_display_name)\n",
    "\n",
    "amp_id_to_diplay_name = {\n",
    "    tpl.amp_id: tpl.display_name\n",
    "    for tpl in bacteriocins_df.itertuples()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Actagardine',\n",
       " 'Bacteriocin II',\n",
       " 'Caulonodin III',\n",
       " 'Colicin M',\n",
       " 'Enterocin 1071B',\n",
       " 'Enterocin NKR-5-3D',\n",
       " 'Enterocin P',\n",
       " 'Garvicin Q',\n",
       " 'Lactococcin A',\n",
       " 'Lactococcin MMFII',\n",
       " 'Lactococcin Q',\n",
       " 'Nocardithiocin',\n",
       " 'Piricyclamide 7005E2',\n",
       " 'Sakacin A'}"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriocins_to_exclude = {\n",
    "    'pg_hydrolase',\n",
    "    'Auto_Inducing_Peptide_III',\n",
    "    'Linocin_M18',\n",
    "    'FlvA2f',\n",
    "    'FlvA2h',\n",
    "    'Competence',\n",
    "    'ComX2',\n",
    "    'ComX4',\n",
    "}\n",
    "bacteriocins_shortlist = set({\n",
    "    amp_id_to_diplay_name[amp_id]\n",
    "    for amp_id in amp_table.index\n",
    "    if amp_id not in bacteriocins_to_exclude\n",
    "})\n",
    "bacteriocins_shortlist"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Selection of bacteriocins present in bacteria only\n",
    "\n",
    "Focus on 4 well-known example as illustration that not all bacteriocins are found in archaea."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_genomes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>display_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Lactococcin 972</th>\n",
       "      <td>273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Closticin 574</th>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UviB</th>\n",
       "      <td>214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Colicin E6</th>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GarvieacinQ garQ</th>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Colicin E9</th>\n",
       "      <td>167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteriocin IIc</th>\n",
       "      <td>161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Microviridin K</th>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ComC</th>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Subtilosin</th>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Colicin V</th>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Avermipeptin</th>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AdDLP</th>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stackepeptin</th>\n",
       "      <td>46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Cyclothiazomycin B</th>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Propionicin SM1</th>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mvan 2782 putative Linaridin</th>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Lactocyclicin Q</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Paeninodin</th>\n",
       "      <td>39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Propeptin</th>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              n_genomes\n",
       "display_name                           \n",
       "Lactococcin 972                     273\n",
       "Closticin 574                       221\n",
       "UviB                                214\n",
       "Colicin E6                          182\n",
       "GarvieacinQ garQ                    172\n",
       "Colicin E9                          167\n",
       "Bacteriocin IIc                     161\n",
       "Microviridin K                       93\n",
       "ComC                                 61\n",
       "Subtilosin                           54\n",
       "Colicin V                            53\n",
       "Avermipeptin                         53\n",
       "AdDLP                                53\n",
       "Stackepeptin                         46\n",
       "Cyclothiazomycin B                   45\n",
       "Propionicin SM1                      44\n",
       "Mvan 2782 putative Linaridin         42\n",
       "Lactocyclicin Q                      39\n",
       "Paeninodin                           39\n",
       "Propeptin                            38"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriocins_df[\n",
    "    ~(bacteriocins_df['amp_id'].isin(shared_bacteriocins))\n",
    "][['display_name', 'assembly_accession']].groupby('display_name').nunique().rename(columns={\n",
    "    'assembly_accession': 'n_genomes'\n",
    "}).sort_values('n_genomes', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacteria_only_shortlist = [\n",
    "    'Lactococcin 972',\n",
    "    'Closticin 574',\n",
    "    'Colicin E6',\n",
    "    'Colicin E9',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of bacteriocins: 18\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Lactococcin A',\n",
       " 'Nocardithiocin',\n",
       " 'Lactococcin Q',\n",
       " 'Enterocin NKR-5-3D',\n",
       " 'Lactococcin MMFII',\n",
       " 'Sakacin A',\n",
       " 'Enterocin P',\n",
       " 'Garvicin Q',\n",
       " 'Bacteriocin II',\n",
       " 'Enterocin 1071B',\n",
       " 'Colicin M',\n",
       " 'Piricyclamide 7005E2',\n",
       " 'Caulonodin III',\n",
       " 'Actagardine',\n",
       " 'Lactococcin 972',\n",
       " 'Closticin 574',\n",
       " 'Colicin E6',\n",
       " 'Colicin E9']"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteriocins_to_plot = [\n",
    "    amp_id_to_diplay_name[amp_id] \n",
    "    for amp_id in amp_table.index\n",
    "    if amp_id not in bacteriocins_to_exclude\n",
    "] + bacteria_only_shortlist\n",
    "\n",
    "print(f'Number of bacteriocins: {len(bacteriocins_to_plot)}')\n",
    "\n",
    "bacteriocins_to_plot"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Make heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_heatmap_data(\n",
    "    gtdb_metadata, \n",
    "    bacteriocins_df, \n",
    "    phylum_groups, \n",
    "    bacteriocin_names, \n",
    "):\n",
    "    gtdb_metadata['phylum_group'] = gtdb_metadata['gtdb_phylum'].apply(lambda p: map_phylm_groups.get(p))\n",
    "    metadata = gtdb_metadata[gtdb_metadata['phylum_group'].isin(phylum_groups)].reset_index()\n",
    "    phylum_count_df = metadata[\n",
    "        ['phylum_group', 'ncbi_accession']\n",
    "    ].groupby('phylum_group').count().rename(columns={\n",
    "        'ncbi_accession': 'n_genomes'\n",
    "    })\n",
    "    phylum_count = {\n",
    "        phylum: phylum_count_df.loc[phylum, 'n_genomes']\n",
    "        for phylum in phylum_groups\n",
    "    }\n",
    "\n",
    "    heatmap_data = {\n",
    "        'phylum_group': [],\n",
    "        'n_assemblies': [],\n",
    "    }\n",
    "    for bacteriocin in bacteriocin_names:\n",
    "        heatmap_data[bacteriocin] = []\n",
    "\n",
    "    for phylum in phylum_groups:\n",
    "        bacteriocin_count = bacteriocins_df[\n",
    "            bacteriocins_df['display_name'].isin(bacteriocin_names) &\n",
    "            (bacteriocins_df['phylum_group'] == phylum)\n",
    "        ][['display_name', 'ncbi_accession']].groupby('display_name').nunique().rename(columns={\n",
    "            'ncbi_accession': 'n_genomes'\n",
    "        })\n",
    "\n",
    "        heatmap_data['phylum_group'].append(phylum)\n",
    "        heatmap_data['n_assemblies'].append(phylum_count[phylum])\n",
    "        for bacteriocin in bacteriocin_names:\n",
    "            if bacteriocin in bacteriocin_count.index:\n",
    "                count = bacteriocin_count.loc[bacteriocin, 'n_genomes']\n",
    "            else:\n",
    "                count = 0\n",
    "            \n",
    "            heatmap_data[bacteriocin].append(count)\n",
    "\n",
    "    return pd.DataFrame.from_dict(heatmap_data).set_index('phylum_group', drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_assemblies</th>\n",
       "      <th>Lactococcin A</th>\n",
       "      <th>Nocardithiocin</th>\n",
       "      <th>Lactococcin Q</th>\n",
       "      <th>Enterocin NKR-5-3D</th>\n",
       "      <th>Lactococcin MMFII</th>\n",
       "      <th>Sakacin A</th>\n",
       "      <th>Enterocin P</th>\n",
       "      <th>Garvicin Q</th>\n",
       "      <th>Bacteriocin II</th>\n",
       "      <th>Enterocin 1071B</th>\n",
       "      <th>Colicin M</th>\n",
       "      <th>Piricyclamide 7005E2</th>\n",
       "      <th>Caulonodin III</th>\n",
       "      <th>Actagardine</th>\n",
       "      <th>Lactococcin 972</th>\n",
       "      <th>Closticin 574</th>\n",
       "      <th>Colicin E6</th>\n",
       "      <th>Colicin E9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phylum_group</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Aenigmatarchaeota</th>\n",
       "      <td>132</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Altiarchaeota</th>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Asgardarchaeota</th>\n",
       "      <td>188</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hadarchaeota</th>\n",
       "      <td>12</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Halobacteriota</th>\n",
       "      <td>739</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hydrothermarchaeota</th>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Iainarchaeota</th>\n",
       "      <td>68</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Methanobacteriota</th>\n",
       "      <td>207</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Micrarchaeota</th>\n",
       "      <td>242</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nanoarchaeota</th>\n",
       "      <td>695</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Nanohaloarchaeota</th>\n",
       "      <td>26</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thermoplasmatota</th>\n",
       "      <td>598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Thermoproteota</th>\n",
       "      <td>705</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Undinarchaeota</th>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     n_assemblies  Lactococcin A  Nocardithiocin  \\\n",
       "phylum_group                                                       \n",
       "Aenigmatarchaeota             132              0               0   \n",
       "Altiarchaeota                  26              0               0   \n",
       "Asgardarchaeota               188              2               0   \n",
       "Hadarchaeota                   12              0               0   \n",
       "Halobacteriota                739              0               4   \n",
       "Hydrothermarchaeota            16              0               0   \n",
       "Iainarchaeota                  68              0               0   \n",
       "Methanobacteriota             207              1               0   \n",
       "Micrarchaeota                 242              1               0   \n",
       "Nanoarchaeota                 695              1               0   \n",
       "Nanohaloarchaeota              26              0               0   \n",
       "Thermoplasmatota              598              0               0   \n",
       "Thermoproteota                705              0               1   \n",
       "Undinarchaeota                  5              0               0   \n",
       "\n",
       "                     Lactococcin Q  Enterocin NKR-5-3D  Lactococcin MMFII  \\\n",
       "phylum_group                                                                \n",
       "Aenigmatarchaeota                0                   0                  0   \n",
       "Altiarchaeota                    0                   0                  0   \n",
       "Asgardarchaeota                  0                   1                  0   \n",
       "Hadarchaeota                     0                   0                  0   \n",
       "Halobacteriota                   1                   0                  1   \n",
       "Hydrothermarchaeota              0                   0                  0   \n",
       "Iainarchaeota                    0                   0                  0   \n",
       "Methanobacteriota                0                   1                  0   \n",
       "Micrarchaeota                    0                   0                  0   \n",
       "Nanoarchaeota                    0                   1                  2   \n",
       "Nanohaloarchaeota                0                   0                  0   \n",
       "Thermoplasmatota                 3                   0                  0   \n",
       "Thermoproteota                   0                   1                  1   \n",
       "Undinarchaeota                   0                   0                  0   \n",
       "\n",
       "                     Sakacin A  Enterocin P  Garvicin Q  Bacteriocin II  \\\n",
       "phylum_group                                                              \n",
       "Aenigmatarchaeota            0            0           0               0   \n",
       "Altiarchaeota                0            0           0               0   \n",
       "Asgardarchaeota              0            0           1               0   \n",
       "Hadarchaeota                 0            0           0               0   \n",
       "Halobacteriota               1            2           0               1   \n",
       "Hydrothermarchaeota          0            0           0               0   \n",
       "Iainarchaeota                1            0           0               0   \n",
       "Methanobacteriota            0            0           0               0   \n",
       "Micrarchaeota                1            1           0               0   \n",
       "Nanoarchaeota                0            0           1               0   \n",
       "Nanohaloarchaeota            0            0           0               0   \n",
       "Thermoplasmatota             0            0           0               1   \n",
       "Thermoproteota               0            1           0               0   \n",
       "Undinarchaeota               0            0           0               0   \n",
       "\n",
       "                     Enterocin 1071B  Colicin M  Piricyclamide 7005E2  \\\n",
       "phylum_group                                                            \n",
       "Aenigmatarchaeota                  1          1                     0   \n",
       "Altiarchaeota                      0          0                     0   \n",
       "Asgardarchaeota                    0          0                     0   \n",
       "Hadarchaeota                       0          0                     0   \n",
       "Halobacteriota                     1          0                     0   \n",
       "Hydrothermarchaeota                0          0                     0   \n",
       "Iainarchaeota                      0          0                     0   \n",
       "Methanobacteriota                  0          0                     0   \n",
       "Micrarchaeota                      0          1                     0   \n",
       "Nanoarchaeota                      0          0                     1   \n",
       "Nanohaloarchaeota                  0          0                     0   \n",
       "Thermoplasmatota                   0          0                     0   \n",
       "Thermoproteota                     0          1                     1   \n",
       "Undinarchaeota                     0          0                     0   \n",
       "\n",
       "                     Caulonodin III  Actagardine  Lactococcin 972  \\\n",
       "phylum_group                                                        \n",
       "Aenigmatarchaeota                 0            0                0   \n",
       "Altiarchaeota                     0            0                0   \n",
       "Asgardarchaeota                   0            0                0   \n",
       "Hadarchaeota                      0            0                0   \n",
       "Halobacteriota                    1            1                0   \n",
       "Hydrothermarchaeota               0            0                0   \n",
       "Iainarchaeota                     0            0                0   \n",
       "Methanobacteriota                 0            0                0   \n",
       "Micrarchaeota                     0            0                0   \n",
       "Nanoarchaeota                     0            1                0   \n",
       "Nanohaloarchaeota                 0            0                0   \n",
       "Thermoplasmatota                  1            0                0   \n",
       "Thermoproteota                    0            0                0   \n",
       "Undinarchaeota                    0            0                0   \n",
       "\n",
       "                     Closticin 574  Colicin E6  Colicin E9  \n",
       "phylum_group                                                \n",
       "Aenigmatarchaeota                0           0           0  \n",
       "Altiarchaeota                    0           0           0  \n",
       "Asgardarchaeota                  0           0           0  \n",
       "Hadarchaeota                     0           0           0  \n",
       "Halobacteriota                   0           0           0  \n",
       "Hydrothermarchaeota              0           0           0  \n",
       "Iainarchaeota                    0           0           0  \n",
       "Methanobacteriota                0           0           0  \n",
       "Micrarchaeota                    0           0           0  \n",
       "Nanoarchaeota                    0           0           0  \n",
       "Nanohaloarchaeota                0           0           0  \n",
       "Thermoplasmatota                 0           0           0  \n",
       "Thermoproteota                   0           0           0  \n",
       "Undinarchaeota                   0           0           0  "
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "archaeal_heatmap_df = compute_heatmap_data(\n",
    "    gtdb_metadata, \n",
    "    bacteriocins_df[bacteriocins_df['domain'] == 'Archaea'].copy(), \n",
    "    phylum_groups=arc_phylum_groups, \n",
    "    bacteriocin_names=bacteriocins_to_plot, \n",
    ")\n",
    "archaeal_heatmap_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n_assemblies</th>\n",
       "      <th>Lactococcin A</th>\n",
       "      <th>Nocardithiocin</th>\n",
       "      <th>Lactococcin Q</th>\n",
       "      <th>Enterocin NKR-5-3D</th>\n",
       "      <th>Lactococcin MMFII</th>\n",
       "      <th>Sakacin A</th>\n",
       "      <th>Enterocin P</th>\n",
       "      <th>Garvicin Q</th>\n",
       "      <th>Bacteriocin II</th>\n",
       "      <th>Enterocin 1071B</th>\n",
       "      <th>Colicin M</th>\n",
       "      <th>Piricyclamide 7005E2</th>\n",
       "      <th>Caulonodin III</th>\n",
       "      <th>Actagardine</th>\n",
       "      <th>Lactococcin 972</th>\n",
       "      <th>Closticin 574</th>\n",
       "      <th>Colicin E6</th>\n",
       "      <th>Colicin E9</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>phylum_group</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Acidobacteriota</th>\n",
       "      <td>1094</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Actinomycetota</th>\n",
       "      <td>4027</td>\n",
       "      <td>11</td>\n",
       "      <td>8</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>7</td>\n",
       "      <td>139</td>\n",
       "      <td>143</td>\n",
       "      <td>17</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aquificota</th>\n",
       "      <td>84</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Armatimonadota</th>\n",
       "      <td>187</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacillota</th>\n",
       "      <td>10923</td>\n",
       "      <td>11</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>15</td>\n",
       "      <td>17</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>131</td>\n",
       "      <td>60</td>\n",
       "      <td>10</td>\n",
       "      <td>17</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 n_assemblies  Lactococcin A  Nocardithiocin  Lactococcin Q  \\\n",
       "phylum_group                                                                  \n",
       "Acidobacteriota          1094              4               2              0   \n",
       "Actinomycetota           4027             11               8              5   \n",
       "Aquificota                 84              0               0              0   \n",
       "Armatimonadota            187              0               0              0   \n",
       "Bacillota               10923             11               3              7   \n",
       "\n",
       "                 Enterocin NKR-5-3D  Lactococcin MMFII  Sakacin A  \\\n",
       "phylum_group                                                        \n",
       "Acidobacteriota                   0                  0          1   \n",
       "Actinomycetota                    2                  0          0   \n",
       "Aquificota                        0                  0          0   \n",
       "Armatimonadota                    0                  0          0   \n",
       "Bacillota                        12                  2          1   \n",
       "\n",
       "                 Enterocin P  Garvicin Q  Bacteriocin II  Enterocin 1071B  \\\n",
       "phylum_group                                                                \n",
       "Acidobacteriota            1           2               0                0   \n",
       "Actinomycetota             2           0               0                2   \n",
       "Aquificota                 0           0               0                0   \n",
       "Armatimonadota             0           0               1                0   \n",
       "Bacillota                  8           3              15               17   \n",
       "\n",
       "                 Colicin M  Piricyclamide 7005E2  Caulonodin III  Actagardine  \\\n",
       "phylum_group                                                                    \n",
       "Acidobacteriota          2                     4               0            0   \n",
       "Actinomycetota           4                     6               2            7   \n",
       "Aquificota               0                     0               0            0   \n",
       "Armatimonadota           0                     0               0            0   \n",
       "Bacillota                3                     0               2            0   \n",
       "\n",
       "                 Lactococcin 972  Closticin 574  Colicin E6  Colicin E9  \n",
       "phylum_group                                                             \n",
       "Acidobacteriota                0              0           1           2  \n",
       "Actinomycetota               139            143          17          11  \n",
       "Aquificota                     0              1           0           0  \n",
       "Armatimonadota                 0              0           1           0  \n",
       "Bacillota                    131             60          10          17  "
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacterial_heatmap_df = compute_heatmap_data(\n",
    "    gtdb_metadata, \n",
    "    bacteriocins_df[bacteriocins_df['domain'] == 'Bacteria'].copy(), \n",
    "    phylum_groups=bac_phylum_groups, \n",
    "    bacteriocin_names=bacteriocins_to_plot, \n",
    ")\n",
    "bacterial_heatmap_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_heatmap_list(heatmap_df, amp_ids):\n",
    "    heatmap_data = []\n",
    "    for phylum in heatmap_df.index:\n",
    "        row = [phylum]\n",
    "        heatmap_row = heatmap_df.loc[phylum]\n",
    "        for amp_id in amp_ids:\n",
    "            c = round(100 * heatmap_row[amp_id] / heatmap_row['n_assemblies'], 3)\n",
    "            row.append(c)\n",
    "\n",
    "        heatmap_data.append(row)\n",
    "\n",
    "    return heatmap_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "archaeal_heatmap_data = make_heatmap_list(archaeal_heatmap_df, bacteriocins_to_plot)\n",
    "\n",
    "itol_heatmap_annotations(\n",
    "        archaeal_heatmap_data,\n",
    "        field_labels=bacteriocins_to_plot,\n",
    "        output_path=data_folder / 'figure4' / 'ar53_phylum_subset.heatmap.txt',\n",
    "        dataset_label='% genomes in archaeal phylum',\n",
    "        color_min='#ffffff',\n",
    "        color_max='#f58231',\n",
    "        min_value=0,\n",
    "        max_value=2,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "bacterial_heatmap_data = make_heatmap_list(bacterial_heatmap_df, bacteriocins_to_plot)\n",
    "\n",
    "itol_heatmap_annotations(\n",
    "        bacterial_heatmap_data,\n",
    "        field_labels=bacteriocins_to_plot,\n",
    "        output_path=data_folder / 'figure4' / 'bac120_phylum_subset.heatmap.txt',\n",
    "        dataset_label='% genomes in bacterial phylum',\n",
    "        color_min='#ffffff',\n",
    "        color_max='#0173b2',\n",
    "        min_value=0,\n",
    "        max_value=4,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "AvB",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
